Increase default API key auth cache size to 25K.

ywangd opened the pull request:
This PR increases default size of the API key auth cache from 10K to 25K
to cover the majority of expected deployment sizes. Also add trace logging
on eviction for the API key auth cache to help identify any cache thrashing 
issue.

ywangd committed:
Increase default API key auth cache size to 25K.

Also add trace logging on eviction for the API key auth cache to help
identify any cache thrashing issue.

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

ywangd committed:
Limit logging to only eviction due to cache size.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -212,6 +213,14 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
                 .setExpireAfterAccess(ttl)
                 .setMaximumWeight(maximumWeight)
+                .removalListener(notification -> {
+                    if (RemovalNotification.RemovalReason.EVICTED == notification.getRemovalReason()) {
+                        if (getApiKeyAuthCache().count() >= maximumWeight) {
```
This is a reasonable approximation to limit the logging to "eviction due to cache size, not ttl".

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -212,6 +213,14 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
+                .removalListener(notification -> {
+                    if (RemovalNotification.RemovalReason.EVICTED == notification.getRemovalReason()) {
+                        if (getApiKeyAuthCache().count() >= maximumWeight) {
+                            logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                    "possibly due to cache size limit", notification.getKey());
```
I don't think trace logging is sufficient.
It will allow us to turn on logging so we can determine whether there is a cache problem, but I don't think it's proactive enough.

Can we look at what it would take to have something that could safely be logged at `warning` instead?

ywangd replied:
Thanks, I think this would be useful since trace logging is very busy. I pushed [some changes](https://github.com/elastic/elasticsearch/pull/74388/commits/725dff3656a867bcaf2fce5571051b792982d56f) trying to add warnings based on eviction rate as a heuristic. I considered the followings in the implementation:

* Regular high rate of eviction due to cache size is a symptom of thrashing
* Assuming 70ms for the expensive hashing, a single core can perform less than 15 hashing per second. This is number is used as the threshold since a single core (smallest node) would be saturated by this rate of eviction. That is, if 15 eviction occurs in a second, it is likely a cache thrashing.
* The warning logging should be throttled. Since agents check in every 10 min, I chose to check the stats every 10 min and issue warnings if possible thrashing is identified.
* The warning logging should be as quickly as possible, it can afford some inaccuracy (use of LongAdder instead of AtomicLong and decrease/increase instead of creating new LongAdder)

Please let me know if you are OK with this approach. I tried to keep the implementation as simplistic as possible. But there are still some level of complexities. I'll add tests for this new piece of code once we agree on the implementation.

tvernum reviewed (approved):
LGTM, but I think we need to work out how to get some warnings in place when the cache is full.

ywangd committed:
Add warn level logging for API key auth cache thrashing.

The purpose is to proactively warn about potential cache thrashing based
on herustic using eviction rate.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -192,6 +195,11 @@
     private volatile long lastExpirationRunMs;
 
+    private static final long EVICTION_CHECKING_INTERVAL = 600_000L; // 600K ms = 10 minutes
+    private static final long EVICTION_WARNING_THRESHOLD = 15 * EVICTION_CHECKING_INTERVAL / 1000; // 15 eviction per sec = 9000 in 10 min
```
We can tweak this later - please don't hold up the PR while we discuss it, but I think this number could be too high.

9,000 in 10 minutes is certainly thrashing, but I would think it's an issue at a much smaller number as well.
If 1,000 items get evicted (due to cache size) over a 10 minute period, isn't that bad?

I don't know what the best heuristic is, but I don't think it needs to be a sustained rate of 15 per second for the full 10 minutes to count as a problem.

ywangd replied:
Yes this threshold is very liberal. A small node is probably tripped over before the warning is issued. I lowered it to 5 min. We can further tweak it. But I also don't want it to be purely based on the smallest node. For a large deployment, assuming 3 nodes each with multiple cores, this threshold might be OKish?

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -212,6 +220,24 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
                 .setExpireAfterAccess(ttl)
                 .setMaximumWeight(maximumWeight)
+                .removalListener(notification -> {
+                    if (RemovalReason.EVICTED == notification.getRemovalReason() && getApiKeyAuthCache().count() >= maximumWeight) {
+                        evictionCounter.increment();
+                        logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                "possibly due to cache size limit", notification.getKey());
+                        final long previousCheck = lastEvictionCheckedAt.get();
+                        final long now = clock.instant().toEpochMilli();
+                        if (now - previousCheck >= EVICTION_CHECKING_INTERVAL && lastEvictionCheckedAt.compareAndSet(previousCheck, now)) {
+                            final long sum = evictionCounter.sum();
+                            evictionCounter.add(-sum); // reset by decrease
+                            if (sum >= EVICTION_WARNING_THRESHOLD) {
+                                logger.warn("Possible thrashing for API key authentication cache, " +
+                                    "[{}] eviction due to cache size within last [{}] seconds",
+                                    sum, EVICTION_CHECKING_INTERVAL / 1000);
+                            }
+                        }
+                    }
```
Nit: I think this block is long enough to justify being a method rather than an in place lambda.

ywangd replied:
I initially didn't extract it because this code needs the local variable `maximumWeight`. This means the extracted method has to return a lambda. Anyway, I went ahead and extracted it. The preamble of this class is already long enough and we can save some lines here.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -212,6 +220,24 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
+                        logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                "possibly due to cache size limit", notification.getKey());
+                        final long previousCheck = lastEvictionCheckedAt.get();
+                        final long now = clock.instant().toEpochMilli();
```
Should this use `System.nanoTime()` instead of epoch time?

ywangd replied:
I thought `clock.instant().toEpochMilli()` would be faster than `System.nanoTime` because I thought the former is only millisecond accuracy thus not requiring extra read from hardware. But it turned out that `clock.instant()` already does the read for nanos. So it should have no performance advantage. I replaced it with `System.nanoTime`.

ywangd committed:
Merge remote-tracking branch 'origin/master' into api-key-cache-size-logging

ywangd committed:
address feedback

ywangd committed:
More tests

tomcallahan commented:
Trace logging needs to be explicitly enabled to catch an issue, and will presumably be high-volume if a user encounters this situation -- I wonder if some kind of counter would be more appropriate?

ywangd commented:
> Trace logging needs to be explicitly enabled to catch an issue, and will presumably be high-volume if a user encounters this situation -- I wonder if some kind of counter would be more appropriate?

@tomcallahan Thanks for the feedback. Tim had a [similar comment](https://github.com/elastic/elasticsearch/pull/74388#discussion_r655866640) and we have since added a `WARN` level logging to more proactively identify potential thrashing issues. The warning message is issued if there are more than 4500 eviction in last 5 min. These values are selected based on how a single CPU core can be saturated by hash computations (15 per second). The message also gets issued no more than once per 5 minutes to control the volume. Frankly, these numbers are still a guess work. We can still tune them if needed before the release is finalised.

tomcallahan commented:
@ywangd Apologies, I had only gone off the PR description. This SGTM, thank you.
